Adds livetime correction to get data

jajmitchell opened the pull request:
Adds livetime correction to get data.

jajmitchell committed:
Adds livetime correction to get_data

Adds livetime correction to get_data

jajmitchell committed:
Adds detector and time masks

Adds detector and time masks

jajmitchell committed:
Merge branch 'main' into adds_livetime_correction_to_get_data

samaloney reviewed (changes requested):
Looks good - needs docs, change log and test/s

jajmitchell committed:
Implements corrections and get_spectrum function

Implements corrections and get_spectrum function

jajmitchell committed:
get_spectrum and bkg_subtract now work

get_spectrum and bkg_subtract now work

samaloney commented:
Add properties for `livetime` and `livefrac` or `deadfrac`

jajmitchell committed:
SRM now functional with wedge grid transmission

SRM now functional with wedge grid transmission

jajmitchell committed:
SRM fully functional

SRM fully functional

jajmitchell committed:
adds spec_obj

jajmitchell committed:
Mostly working

jajmitchell committed:
Livetime works!!!!

jajmitchell committed:
bkg_sub works

jajmitchell committed:
errors work

jajmitchell committed:
works

jajmitchell committed:
bkg sub finally works!

jajmitchell committed:
Adds systematic errors

jajmitchell committed:
Adds time range to meta in spec object

jajmitchell committed:
removes print statements

samaloney commented on stixpy/calibration/grid.py:
```diff
@@ -7,13 +7,14 @@
 import astropy.units as u
 import numpy as np
 from astropy.table import Table
+import xraydb
```
So we already use `roentgen` should either migrate to `xraydb` or keep using

samaloney commented on stixpy/calibration/grid.py:
```diff
@@ -32,18 +33,141 @@ def get_grid_transmission(flare_location: STIXImaging):
+
+def stx_grid_transmission(pitch, slit, thickness, L):
+
+    ds = 5e-3
+    dh = 5e-2
```
Magic numbers need at least comment and ideally units

samaloney commented on stixpy/calibration/grid.py:
```diff
@@ -32,18 +33,141 @@ def get_grid_transmission(flare_location: STIXImaging):
+    return subc_transm
+
+
+def stx_grid_transmission(pitch, slit, thickness, L):
```
Needs a docstring and test

samaloney commented on stixpy/calibration/grid.py:
```diff
@@ -32,18 +33,141 @@ def get_grid_transmission(flare_location: STIXImaging):
+
+    # ;; Transmission for a wedge shape model for grid imperfections
+    g0 = slit_rep / pitch_rep + (pitch_rep - slit_rep) / pitch_rep * np.exp( - H_rep / L_rep )
+    ttt = L_rep / dh * ( 1. - np.exp(- dh / L_rep ) )
```
naming

jajmitchell replied:
ask paolo

samaloney commented on stixpy/calibration/livetime.py:
```diff
@@ -93,7 +93,7 @@ def pileup_correction_factor():
 
 
-def get_livetime_fraction(trigger_rate, *, eta=1.10 * u.us, tau=10.1 * u.us):
+def get_livetime_fraction(trigger_rate,*, eta=1.1e-6 *u.s, tau=10.1e-6 * u.s):
```
Just wonder what motivated this change?

jajmitchell replied:
change back

samaloney commented on stixpy/io/readers.py:
```diff
@@ -134,8 +134,19 @@ def read_elut(elut_file, sci_channels):
+            elut.offset = elut_table["Offset (ADC)"].reshape(32, 12)
+            elut.gain = elut_table["Gain (ADC/keV)"].reshape(32, 12)
+
+
```
This is already fix on main 

https://github.com/TCDSolar/stixpy/blob/4f4e0c7d7d14ad71ebcf1a2b3882327ad926cc6d/stixpy/io/readers.py#L136-L140

samaloney commented on stixpy/product/sources/science.py:
```diff
@@ -999,15 +1084,529 @@ def get_data(
+        return counts_corr, counts_err, times, t_norm, livefrac, energies, elut_cor_fac
+        
+    
+    def get_spectrum(self, bkg_prod=None):
```
So ideally this should be handled by `get_data` e.g.

```
get_data(detector_indices=[[0,31]], pixel_indices=[[0,8]], elut=True, live=True, ...)
```

samaloney commented on stixpy/product/sources/science.py:
```diff
@@ -999,15 +1084,529 @@ def get_data(
+
+# NEED TO ADD IN THE DIST AND ANGLE READING AND BKG_SUBTRACT, GETTING THERE THOUGH!!!
+
+    def bkg_subtract(self, bkg_data):
```
I think this might be better as a plain function, e.g. not a method on the object.

samaloney commented on stixpy/product/sources/science.py:
```diff
@@ -999,15 +1084,529 @@ def get_data(
+        
+        return spec_sub
+
+    def get_masked_srm(self,flare_location):
```
Need to support flat spectral assumption and slope septette the arm sliceing and other stuff

samaloney reviewed (commented):
Overall this looks great 🚀

Leaving all my comment but I know it still a WIP so some of them I'm sure you'd address anyway.

Few high level things are:
* docstrings
* test
* dockets where it makes sense
* needs a rebase/merge of current main

To me the key  aspect is understanding want someone reading the code to know what's happening and why can link to paper / ref where applicable.

IDListis just because it done one way in IDL doesn't mean we have to or even should do the same - we need to support it for comparisons but they only to a point.

Also need to maintain some level of backwards compatibly with old transmission correction etc and also ideally support incremental correction e.g new transmission but no spectral slope correction. 

As much of this as possible should be done it `get_data` we can create  return a dictionary, dataclass etc. If we wanted to go hole hog `get_data` would return `Spectrum`, `SpectrumCollection`, or `SpectrumSequences` objects but that could be a later PR.

Also I think the u, v coords may need to be updated too with the new transmission?

samaloney committed:
Merge branch 'main' into adds_livetime_correction_to_get_data

jajmitchell committed:
New structure, counts match old version

Everything has been moved onto get_data()

jajmitchell committed:
Multiple features added

Flare location added from Laura's flarelist and slightly modified, flare angle calculator also added, background subration done for each pixel and detector without collapsing

jajmitchell committed:
Functions with data expanded

jajmitchell committed:
Uncertainties working

Commit:
remove print

jajmitchell committed:
fixes SRM energy slicing

jajmitchell committed:
updates submodule

jajmitchell committed:
reads drm fits file

jajmitchell committed:
Fixes uncertainties

jajmitchell committed:
slicing and plotting now functional

jajmitchell committed:
error bar plotting works

jajmitchell committed:
removes event_time_range all functions through time_indices with rcr check

jajmitchell committed:
sunkit-spex detector and pixel indices dimension check

jajmitchell committed:
soft coded srm detector and pixel indices

jajmitchell committed:
NDsequence and NDcollection now functional

jajmitchell committed:
corrects indexing error

jajmitchell committed:
tidy up

jajmitchell committed:
adds doc strings

jajmitchell committed:
adds in energy slicing using quantities rather than just indices

jajmitchell committed:
tidy up

jajmitchell committed:
differing rcr states now added to SRM

jajmitchell committed:
rcr states now work

jajmitchell committed:
some fixes

jajmitchell committed:
fixes rcr shift

jajmitchell committed:
adds rcr docstring

jajmitchell committed:
spec file functions with bkg sub

jajmitchell committed:
sunkit-spex spectrum object for spc file works, still need to add correct grid transmission

jajmitchell committed:
spec srm functions

jajmitchell committed:
fix spectrum file triggers

jajmitchell committed:
Counts match IDL method for detector summation

jajmitchell committed:
errors functional

jajmitchell committed:
1d spec functions plus adds tests

jajmitchell committed:
Nesting should correctly work for bkgsub and systematic error

jajmitchell committed:
no_bkg sub now works and tests have been added

jajmitchell committed:
tidy slightly

jajmitchell committed:
ELUT correction now works for detector sum or independent

jajmitchell committed:
bkg counts and subc match IDL
